Add database

Eugenpak opened the pull request:
Add database. Sprint-12

Eugenpak committed:
Добавить в пакет controller: GenreController, MpaController. Сохранить изменения.

Eugenpak committed:
Добавить в пакет model: Genre, Mpa, PopularFilm. Сохранить изменения.

Eugenpak committed:
Добавить в пакет service: GenreService, MpaService. Сохранить изменения.

Eugenpak committed:
Добавить в пакет storage.film: FilmDbStorage, FilmRowMapper. Сохранить изменения.

Eugenpak committed:
Добавить в пакет storage.filmgenre: FilmGenreDao, FilmGenreDaoImpl.

Eugenpak committed:
Добавить в пакет storage.filmgenre: FilmMpaDao, FilmMpaDaoImpl, FilmMpaRowMapper.

Eugenpak committed:
Добавить в пакет storage.friend: FriendDbStorage, FriendStorage.

Eugenpak committed:
Добавить в пакет storage.genre: GenreDbStorage, GenreStorage, GenreRowMapper.

Eugenpak committed:
Добавить в пакет storage.like: LikeDao, LikeDaoImpl, PopularFilmRowMapper.

Eugenpak committed:
Добавить в пакет storage.mpa: MpaDbStorage, MpaStorage, MpaRowMapper.

Eugenpak committed:
Добавить в пакет storage.user: UserDbStorage, UserRowMapper. Сохранить изменения.

Eugenpak committed:
Добавить в пакет storage: BaseDbStorage.

Eugenpak committed:
Добавить в пакет exception: InternalServerException.

Eugenpak committed:
Добавить в пакет main.resources: InternalServerException.

Eugenpak committed:
Добавить в пакет test.

Eugenpak committed:
Сохранить: pom.xml, README.md.

Eugenpak committed:
Checkstyle Spr-12-FZ-v-1

Eugenpak committed:
Checkstyle Spr-12-FZ-v-2

Eugenpak committed:
Checkstyle Spr-12-FZ-v-3

Eugenpak committed:
Checkstyle Spr-12-FZ-v-4

Eugenpak committed:
Checkstyle Spr-12-FZ-v-5

Eugenpak committed:
Checkstyle Spr-12-FZ-v-6

Eugenpak committed:
Checkstyle Spr-12-FZ-v-7

Eugenpak committed:
Checkstyle Spr-12-FZ-v-8

Eugenpak committed:
Checkstyle Spr-12-FZ-v-9

Eugenpak committed:
Checkstyle Spr-12-FZ-v-10

Eugenpak committed:
Checkstyle Spr-12-FZ-v-11

Eugenpak committed:
Checkstyle Spr-12-FZ-v-12

Eugenpak committed:
Checkstyle Spr-12-FZ-v-13

Eugenpak committed:
Сохранить тесты.

Eugenpak committed:
Checkstyle  Spr-12-FZ-v-14

Eugenpak committed:
Сохранить изменение в пакете storage\film.

Eugenpak committed:
Сохранить изменение в пакете storage\friend.

Eugenpak committed:
Сохранить изменение в пакете storage\like.

Eugenpak committed:
Сохранить изменение в пакете storage\user.

Eugenpak committed:
Сохранить изменения в тестах в пакете test\...\service.

Eugenpak committed:
Сохранить изменения в тестах в пакете test\...\storage.

Eugenpak committed:
Checkstyle Spr-12-FZ-v-15

Eugenpak committed:
Checkstyle Spr-12-FZ-v-16

Eugenpak committed:
Checkstyle Spr-12-FZ-v-17

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -27,6 +27,12 @@ public FilmController(FilmService filmService) {
     @GetMapping
     public Collection<Film> findAll() {
         log.info("Start Film findAll()");
+        //return filmService.findAll();
```
Если код больше не нужен, его лучше удалить.

Eugenpak replied:
Принято. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,39 @@
+    public Genre findGenreById(@NotNull @PathVariable long id) {
+        log.info("GC Genre findGenreById(id=" + id + ")");
+        Genre findGenre = genreService.findGenreById(id);
+        //System.out.println("GC findGenreById= " + findGenre);
```
Здесь и дальше, после окончания разработки отладочную печать лучше удалить

Eugenpak replied:
Принято. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -1,12 +1,14 @@
+@Component
+@Qualifier("inMemoryUserStorage")
 @Slf4j
 public class InMemoryUserStorage implements UserStorage {
```
На Ваше усмотрение. 
Раз перешли на работу с базой данных, то смысла в этих репозитариях, в общем-то, нет, их можно спокойно удалить из проекта.

Eugenpak replied:
Принято. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/storage/BaseDbStorage.java:
```diff
@@ -0,0 +1,66 @@
+            for (int idx = 0; idx < params.length; idx++) {
+                ps.setObject(idx + 1, params[idx]);
+            }
+            System.out.println(ps);
```
Потом лучше удалить отладочную печать

Eugenpak replied:
Принято. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,173 @@
+    private final FilmGenreDao filmGenreDao;
+    private final MpaStorage mpaStorage;
+    private final GenreStorage genreStorage;
+    private final LikeDao likeDao;
```
```java
private final FilmMpaDao filmMpaDao;
    private final FilmGenreDao filmGenreDao;
    private final MpaStorage mpaStorage;
    private final GenreStorage genreStorage;
    private final LikeDao likeDao;
```
Вообще говоря, нет, это нехорошо. Репозитории (DAO) ничего друг про друга знать не должны. Репозиторий выполняет  crud команды для своей таблицы (набора данных) и,  если надо,  возвращает результаты в сервис. 
Если же репозиторий вызывает другие репозитории, то есть фактически зависит от них, то это серьёзный повод пересмотреть архитектуру программы. 
Вся бизнес логика находится в сервисах. Сервис может обращаться к другим сервисам, а также получать данные из нескольких репозиториев. Репозиторий же настроен исключительно на свою таблицу.

Eugenpak replied:
Принято. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,173 @@
+    private final MpaStorage mpaStorage;
+    private final GenreStorage genreStorage;
+    private final LikeDao likeDao;
+
```
Вся работа по сборке и обработке данных должна происходить в сервисе. Потому что это логика программы, то есть бизнес-правила, а сервисы как раз и предназначены для бизнес-правил.
А в репозитариях лучше оставить только базовые операции с соответствующей таблицей (набором данных).

Eugenpak replied:
Принято. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,173 @@
+        log.info("FilmDbStorage start findAll()");
+        Collection<Film> listFilm = findMany(FIND_ALL_QUERY);
+        for (Film el : listFilm) {
+            el.setGenres(filmGenreDao.findGenresById(el.getId()));
```
В этом месте, в цикле,  идут запросы к базе данных. Хорошо бы переделать данный метод, чтобы уменьшить число обращений к БД .

Eugenpak replied:
Принято. Исправил.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/storage/filmgenre/FilmGenreDaoImpl.java:
```diff
@@ -0,0 +1,81 @@
+    public void addSet(long filmId, Set<Genre> genres) {
+        log.debug("FilmGenreDaoImpl addSet({}, {}).", filmId, genres);
+        for (Genre el: genres) {
+            int rowsAdd = jdbcTemplate.update(INSERT_QUERY, filmId, el.getId());
```
В этом месте, в цикле, всё время дёргается база данных. Было бы хорошо избавиться от этих многочисленных запросов, которые сильно нагружают сервер. Хорошо было бы  сохранить данные одним запросом, например, с помощью конструкции:
jdbcTemplate.batchUpdate(............

Eugenpak replied:
Принято. Исправил.

avfyodorov reviewed (commented):
Добрый день, Евгений! 

В целом хорошо, правда, есть ряд уточнений. 
К тестам вопросов нет, тесты хорошие получились.

Eugenpak committed:
Сохранить пакет model.

Eugenpak committed:
Сохранить пакет service.

Eugenpak committed:
Сохранить пакет film, filmgenre, filmmpa, genre, mpa, user. Исправить ошибки.
https://github.com/Eugenpak/java-filmorate/pull/3#discussion_r2022529083
https://github.com/Eugenpak/java-filmorate/pull/3#discussion_r2025815851
https://github.com/Eugenpak/java-filmorate/pull/3#discussion_r2022524664
https://github.com/Eugenpak/java-filmorate/pull/3#discussion_r2022517182

Eugenpak committed:
Сохранить изменения тесты.

Eugenpak committed:
Checkstyle Spr-12-FZ-v-18

Eugenpak committed:
Checkstyle Spr-12-FZ-v-19

Eugenpak committed:
Checkstyle Spr-12-FZ-v-20

Eugenpak committed:
Отчистка кода.

Eugenpak committed:
Сохранить изменения в пакете service: FilmService.

Eugenpak committed:
Checkstyle Spr-12-FZ-v-20

avfyodorov reviewed (commented):
Добрый день, Евгений!

Замечаний нет.
Работа принята.
